src/cio_websocket.c

```diff
@@ -0,0 +1,466 @@
+
+	memcpy(ws->mask, ptr, sizeof(ws->mask));
+	if (likely(ws->read_frame_length > 0)) {
+		bs->read_exactly(bs, buffer, ws->read_frame_length, get_payload, ws);
+	} else {
+		buffer->bytes_transferred = 0;
+		handle_frame(ws, NULL, 0);
+	}
+}
+
+static void get_mask_or_payload(struct cio_websocket *ws, struct cio_buffered_stream *bs, struct cio_read_buffer *buffer)
+{
+	if (ws->ws_flags.mask == 1) {
+		bs->read_exactly(bs, buffer, sizeof(ws->mask), get_mask, ws);
+	} else {
+		if (likely(ws->read_frame_length > 0)) {
+			bs->read_exactly(bs, buffer, ws->read_frame_length, get_payload, ws);
+		} else {
+			handle_frame(ws, NULL, 0);
+		}
+	}
+}
+
+static void get_length16(struct cio_buffered_stream *bs, void *handler_context, enum cio_error err, struct cio_read_buffer *buffer)
+{
+	size_t len = cio_read_buffer_get_transferred_bytes(buffer);
+	if (unlikely((err != CIO_SUCCESS) || (len == 0))) {
+		// TODO: fill out
+		return;
+	}
+
+	uint8_t *ptr = cio_read_buffer_get_read_ptr(buffer);
+	struct cio_websocket *ws = (struct cio_websocket *)handler_context;
+
+	uint16_t field;
+	memcpy(&field, ptr, sizeof(field));
+	field = cio_be16toh(field);
+	ws->read_frame_length = (uint64_t)field;
+	get_mask_or_payload(ws, bs, buffer);
+}
+
+static void get_length64(struct cio_buffered_stream *bs, void *handler_context, enum cio_error err, struct cio_read_buffer *buffer)
+{
+	size_t len = cio_read_buffer_get_transferred_bytes(buffer);
+	if (unlikely((err != CIO_SUCCESS) || (len == 0))) {
+		// TODO: fill out
+		return;
+	}
+
+	uint8_t *ptr = cio_read_buffer_get_read_ptr(buffer);
+	struct cio_websocket *ws = (struct cio_websocket *)handler_context;
+
+	uint64_t field;
+	memcpy(&field, ptr, sizeof(field));
+	field = cio_be64toh(field);
+	ws->read_frame_length = field;
+	get_mask_or_payload(ws, bs, buffer);
+}
+
+static void get_first_length(struct cio_buffered_stream *bs, void *handler_context, enum cio_error err, struct cio_read_buffer *buffer)
+{
+	size_t len = cio_read_buffer_get_transferred_bytes(buffer);
+	if (unlikely((err != CIO_SUCCESS) || (len == 0))) {
+		// TODO: fill out
+		return;
+	}
+
+	uint8_t *ptr = cio_read_buffer_get_read_ptr(buffer);
+	uint8_t field = *ptr;
+	struct cio_websocket *ws = (struct cio_websocket *)handler_context;
+
+	if ((field & WS_MASK_SET) == WS_MASK_SET) {
+		ws->ws_flags.mask = 1;
+	} else {
+		ws->ws_flags.mask = 0;
+	}
+
+	field = field & ~WS_MASK_SET;
+	if (field < WS_SMALL_FRAME_SIZE + 1) {
+		ws->read_frame_length = (uint64_t)field;
+		get_mask_or_payload(ws, bs, buffer);
+	} else if (field == WS_SMALL_FRAME_SIZE + 1) {
+		bs->read_exactly(bs, buffer, 2, get_length16, ws);
+	} else {
+		bs->read_exactly(bs, buffer, 8, get_length64, ws);
+	}
+}
+
+static void get_header(struct cio_buffered_stream *bs, void *handler_context, enum cio_error err, struct cio_read_buffer *buffer)
+{
+	size_t len = cio_read_buffer_get_transferred_bytes(buffer);
+	if (unlikely((err != CIO_SUCCESS) || (len == 0))) {
+		// TODO: fill out
```

Comment:
![INFO](https://sonarsource.github.io/sonar-github/severity-info.png 'Severity: INFO') Complete the task associated to this "TODO" comment. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](https://sonarcloud.io/coding_rules#rule_key=c%3AS1135)